File: 2024/day_3/solution.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"regexp"
	"strconv"
)

func Solve(fname string) {
	instruction := ReadInput(fname)
	solution1 := Part1(instruction)
	solution2 := Part2(instruction)

	fmt.Printf("Part 1: %d\nPart 2: %d\n", solution1, solution2)
}

func find_mult(instruction string) int {
	re := regexp.MustCompile(`mul\(\d{1,5},\d{1,5}\)`)
	matches := re.FindAllString(instruction, -1)

	re_numbers := regexp.MustCompile(`\d+`)

	var total int
	for _, match := range matches {
		// fmt.Println(match)
		numbers := re_numbers.FindAllString(match, -1)
		mul := 1
		for _, number := range numbers {
			num, _ := strconv.Atoi(number)
			mul *= num
		}
		total += mul
	}
	return total
}

func Part1(instruction string) int {
	total := find_mult(instruction)
	return total
}

func Part2(instruction string) int {
	safe_report_counter := 0

	re := regexp.MustCompile(`(do\(\)|don't\(\))`)
	matches := re.FindAllStringIndex(instruction, -1)
	last_index := 0
	consider := true

	for _, match := range matches {
		last_text := instruction[last_index:match[0]]
		next_instruction := instruction[match[0]:match[1]]
		fmt.Println(last_index, match[0], next_instruction)

		if consider {
			// fmt.Println(last_text)
			// fmt.Println()
			current := find_mult(last_text)
			safe_report_counter += current
		}
		if next_instruction == "do()" {
			consider = true
		} else {
			consider = false
		}
		last_index = match[1]
	}

	return safe_report_counter
}

func ReadInput(fname string) string {
	file, err := os.Open(fname)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	scan := bufio.NewScanner(file)

	var full_string string
	for scan.Scan() {
		full_string += scan.Text()
	}

	return full_string
}

func main() {
	path, _ := os.Getwd()
	filepath := path + "/input.txt"
	Solve(filepath)
}

File: 2024/day_3/solution.py
import re


def read():
    with open("input.txt") as f:
        return f.read()

PATTERN=r"mul\(\d{1,5},\d{1,5}\)"

def check_report(report):
    last_diff = report[1] - report[0]
    if abs(last_diff) > 3 or abs(last_diff) < 1:
        return 0
    for i in range(len(report[2:])):
        diff = report[i+2] - report[i+1]
        if diff*last_diff < 0 or abs(diff) > 3 or abs(diff) < 1:
            return 0
        last_diff = diff
    return 1

if __name__ == "__main__":
    text = read()
    first_list = []
    second_list = []

    reports = []

    total = 0
    matches = re.findall(PATTERN, text)
    for m in matches:
        numbers = m.lstrip('mul(').rstrip(')').split(',')
        mul = 1
        for number in numbers:
            mul *= int(number)
        total += mul
    print(total)

    total = 0
    sections = re.split(r"(do\(\)|don't\(\))", text)
    consider = True
    for s in sections:
        if s == "don't()":
            consider = False
        elif s == 'do()':
            consider = True
        else:
            if not consider:
                continue
            matches = re.findall(PATTERN, s)
            for m in matches:
                numbers = m.lstrip('mul(').rstrip(')').split(',')
                mul = 1
                for number in numbers:
                    mul *= int(number)
                # print(mul)
                total += mul
        # print(numbers, mul)
    print(total)
